File: web/pruefer.js
/**
 * Regel-Pruefer der Kaltakquise-Maschine, Browser-Fassung.
 *
 * Portierung von src/kaltakquise/pruefer.py. Gleiche Regeln, gleiche
 * Meldungen, gleiche Schweregrade. Aendert sich das Regelwerk, gehoert die
 * Aenderung in beide Dateien; tests/test_web_pruefer.mjs haelt sie deckungsgleich.
 */

export const FEHLER = "fehler";
export const WARNUNG = "warnung";
export const HINWEIS = "hinweis";

const RANG = { [FEHLER]: 0, [WARNUNG]: 1, [HINWEIS]: 2 };

export const KI_MARKER = [
  "ich hoffe, diese e-mail erreicht sie gut",
  "ich hoffe, diese email erreicht sie gut",
  "in der heutigen schnelllebigen zeit",
  "massgeschneiderte lösungen",
  "maßgeschneiderte lösungen",
  "innovativ",
  "ganzheitlich",
  "synergien",
  "mehrwert",
  "gerne stehe ich für rückfragen zur verfügung",
  "zögern sie nicht",
  "revolutionär",
  "einzigartige gelegenheit",
  "ich wollte mich kurz vorstellen",
  "lassen sie uns gemeinsam",
  "auf augenhöhe",
  "am ende des tages",
  "wir freuen uns auf ihre rückmeldung",
  "als ihr verlässlicher partner",
  "spannend",
  "spannende",
  "spannendes",
];

export const SIGNATUR = [
  "Burak Ücöz",
  "Sales Engineer Power",
  "Mobil in Time AG, An Aggreko Company",
  "+41 44 806 13 19",
  "burak.ucoez@mobilintime.ch",
];

export const GRUSSFORMELN = {
  ch: "Freundliche Grüsse",
  de: "Mit freundlichen Grüßen",
  at: "Mit freundlichen Grüßen",
  fr: "Meilleures salutations",
  it: "Cordiali saluti",
  en: "Kind regards",
};

export const REGIONEN = {
  ch: "Deutschschweiz, Hochdeutsch, Sie-Form, ss statt ß",
  de: "Deutschland, Hochdeutsch, Sie-Form, ß nach deutscher Rechtschreibung",
  at: "Österreich, Hochdeutsch, Sie-Form, ß nach deutscher Rechtschreibung",
  fr: "Romandie, Französisch, vous-Form",
  it: "Tessin, Italienisch, Lei-Form",
  en: "Internationaler Empfänger, englische Fassung",
};

const ANREDE_MUSTER =
  /^\s*(guten tag|sehr geehrte[rs]?|bonjour|madame|monsieur|gentile|egregio|dear)\b/i;
const ANREDE_VERBOTEN = "sehr geehrte damen und herren";
const GRUSS_MUSTER =
  /^\s*(freundliche gr(ü|ue)sse|mit freundlichen gr(ü|ue|u)(ss|ß)en|meilleures salutations|cordiali saluti|kind regards|best regards)\s*$/i;
const BETREFF_MUSTER = /^\s*(betreff|subject|objet|oggetto)\s*:\s*(.+)/i;

const BETREFF_MAX_ZEICHEN = 50;
const BETREFF_MIN_WOERTER = 4;
const BETREFF_MAX_WOERTER = 8;
const BETREFF_VERBOTEN = ["angebot", "aw:", "re:", "fwd:", "wg:"];

export const FLIESSTEXT_MIN = 90;
export const FLIESSTEXT_MAX = 150;

const PREIS_MUSTER =
  /(chf|eur|€|fr\.|franken|euro)\s*[\d'’.,]+|[\d'’.,]+\s*(chf|eur|€|franken|euro)|\bpro\s+tag\s+(chf|eur)|\btagesmiete\s+(von|ab|chf|eur)/gi;
const VERFUEGBARKEIT_MUSTER =
  /reservier\w*|\bKW\s?\d{1,2}\b|geblockt|halten wir für sie frei|garantier\w*/i;
const VERFUEGBARKEIT_ERLAUBT = /kläre ich|klaere ich|innert 24|innerhalb von 24/i;
const SUPERLATIVE =
  /\b(beste[rsn]?|führend\w*|fuehrend\w*|weltweit|marktführer\w*|marktfuehrer\w*|modernste[rsn]?|optimal\w*|perfekt\w*|einzigartig\w*)\b/gi;

const NORM_IEC_VOLL = /IEC\s?61000-4-30\s+Klasse\s+A/;
const NORM_IEC_KURZ = /IEC\s?61000-4-30/;
const NORM_KLASSE_A = /\bKlasse\s+A\b/;

const EINHEIT_FALSCH = /\b(KVA|Kva|kva|KWH|Kwh|kwh|kwH|MVa|mva|Mw|mw)\b|\bKW\b(?!\s?\d)/g;

const KUERZEL = {
  BESS: "Batteriespeicher",
  USV: "unterbrechungsfreie",
  NEA: "Netzersatzanlage",
};

const CTA_SIGNALE =
  /\?|ein wort genügt|ein wort genuegt|melde ich mich|schicke ich ihnen|rufe ich sie|passt ihnen|hätten sie|haetten sie/i;
const CTA_SCHWACH = /bei interesse|melden sie sich gerne|freue mich auf ihre antwort/i;

const GEDANKENSTRICH = /[–—]/;
const BULLET_ZEILE = /^\s*([-*•‣]|\d+[.)])\s+/;
const MARKDOWN_MUSTER = /\*\*|__|^#{1,6}\s/m;
const LUECKE_MUSTER = /\[[^\]]{2,}\]/g;

const EMOJI_BEREICHE = [
  [0x2600, 0x27bf],
  [0x2190, 0x21ff],
  [0x2b00, 0x2bff],
];

const STOPP_WOERTER = new Set([
  "eine", "einer", "einem", "einen", "sich", "sind", "wird", "werden", "haben",
  "diese", "dieser", "dieses", "nicht", "auch", "aber", "oder", "wenn", "dann",
  "noch", "schon", "sehr", "mehr", "kann", "muss", "soll", "dass", "über",
  "unter", "nach", "beim", "vom", "zum", "zur", "durch", "gegen", "firma",
  "kunde", "kunden", "neue", "neuer", "neues", "grosse", "grosser",
]);

const istWort = (w) => /[\p{L}\p{N}]/u.test(w);
const woerter = (text) => text.split(/\s+/).filter((w) => w && istWort(w));
const normalisiert = (text) => text.toLowerCase().replace(/\s+/g, " ");

function istEmoji(zeichen) {
  if (/\p{Extended_Pictographic}/u.test(zeichen)) return true;
  const code = zeichen.codePointAt(0);
  return EMOJI_BEREICHE.some(([a, b]) => code >= a && code <= b);
}

/** Zerlegt einen Mailtext. Fehlende Teile bleiben leer, der Prüfer meldet sie. */
export function zerlege(text) {
  const roh = String(text).replace(/\r\n/g, "\n").trim();
  const zeilen = roh.split("\n");

  let betreff = "";
  let start = 0;
  for (let i = 0; i < Math.min(5, zeilen.length); i++) {
    const treffer = zeilen[i].match(BETREFF_MUSTER);
    if (treffer) {
      betreff = treffer[2].trim().replace(/^["„“]|["„“]$/g, "");
      start = i + 1;
      break;
    }
  }

  let anrede = "";
  let anredeIdx = start;
  for (let i = start; i < zeilen.length; i++) {
    if (zeilen[i].trim()) {
      anredeIdx = i;
      if (ANREDE_MUSTER.test(zeilen[i])) anrede = zeilen[i].trim();
      break;
    }
  }

  let gruss = "";
  let grussIdx = zeilen.length;
  for (let i = zeilen.length - 1; i > anredeIdx; i--) {
    if (GRUSS_MUSTER.test(zeilen[i])) {
      gruss = zeilen[i].trim();
      grussIdx = i;
      break;
    }
  }

  const korpusStart = anrede ? anredeIdx + 1 : anredeIdx;
  const fliesstext = zeilen.slice(korpusStart, grussIdx).join("\n").trim();
  const signatur = zeilen.slice(grussIdx + 1).map((z) => z.trim()).filter(Boolean);

  return {
    roh,
    betreff,
    anrede,
    fliesstext,
    gruss,
    signatur,
    wortzahl: woerter(fliesstext).length,
  };
}

function signalwoerter(quelle) {
  const gefunden = quelle.match(/[\p{L}\p{N}.\-]{3,}/gu) || [];
  const signale = [];
  for (const wort of gefunden) {
    const rein = wort.replace(/^[.\-]+|[.\-]+$/g, "");
    if (!rein || STOPP_WOERTER.has(rein.toLowerCase())) continue;
    const hatZiffer = /\d/.test(rein);
    const istName = rein[0] === rein[0].toUpperCase() && /\p{L}/u.test(rein[0]) && rein.length >= 4;
    if (hatZiffer || istName) signale.push(rein);
  }
  return [...new Set(signale)];
}

/** Prüft eine Mail und liefert alle Befunde, schwerste zuerst. */
export function pruefe(text, { region = "ch", quelle = null } = {}) {
  const mail = typeof text === "string" ? zerlege(text) : text;
  const befunde = [];
  const melde = (schwere, regel, meldung) => befunde.push({ schwere, regel, text: meldung });

  const roh = mail.roh;
  const flach = normalisiert(roh);

  // --- Betreff -----------------------------------------------------------
  if (!mail.betreff) {
    melde(FEHLER, "Betreff", "Kein Betreff gefunden (erste Zeile 'Betreff: ...').");
  } else {
    const anzahl = mail.betreff.split(/\s+/).filter(Boolean).length;
    if (mail.betreff.length > BETREFF_MAX_ZEICHEN) {
      melde(WARNUNG, "Betreff",
        `${mail.betreff.length} Zeichen, maximal ${BETREFF_MAX_ZEICHEN} (sonst auf dem Handy abgeschnitten).`);
    }
    if (anzahl < BETREFF_MIN_WOERTER || anzahl > BETREFF_MAX_WOERTER) {
      melde(WARNUNG, "Betreff",
        `${anzahl} Wörter, vorgesehen sind ${BETREFF_MIN_WOERTER} bis ${BETREFF_MAX_WOERTER}.`);
    }
    if (mail.betreff.includes("?") || mail.betreff.includes("!")) {
      melde(FEHLER, "Betreff", "Kein Frage- oder Ausrufezeichen im Betreff.");
    }
    for (const verboten of BETREFF_VERBOTEN) {
      if (mail.betreff.toLowerCase().includes(verboten)) {
        melde(FEHLER, "Betreff", `Verbotenes Wort im Betreff: '${verboten}'.`);
      }
    }
  }

  // --- Anrede ------------------------------------------------------------
  if (!mail.anrede) melde(FEHLER, "Anrede", "Keine Anrede gefunden.");
  if (flach.includes(ANREDE_VERBOTEN)) {
    melde(FEHLER, "Anrede",
      "'Sehr geehrte Damen und Herren' ist in der Kaltakquise verboten, Namen recherchieren oder Lücke markieren.");
  }

  // --- Einstieg ----------------------------------------------------------
  const erstesWort = (mail.fliesstext.split(/\s+/)[0] || "").replace(/[,.:;]+$/, "");
  if (erstesWort.toLowerCase() === "ich") {
    melde(FEHLER, "Einstieg", "Der erste Satz darf nicht mit 'Ich' beginnen.");
  }

  // --- Länge -------------------------------------------------------------
  if (!mail.fliesstext) {
    melde(FEHLER, "Fliesstext", "Kein Fliesstext gefunden.");
  } else if (mail.wortzahl < FLIESSTEXT_MIN) {
    melde(WARNUNG, "Länge",
      `${mail.wortzahl} Wörter Fliesstext, Ziel sind ${FLIESSTEXT_MIN} bis ${FLIESSTEXT_MAX}.`);
  } else if (mail.wortzahl > FLIESSTEXT_MAX) {
    melde(FEHLER, "Länge",
      `${mail.wortzahl} Wörter Fliesstext, Maximum ist ${FLIESSTEXT_MAX}. Kalt heisst kurz.`);
  }

  // --- Grussformel und Signatur -----------------------------------------
  if (!mail.gruss) {
    melde(FEHLER, "Grussformel", "Keine Grussformel gefunden.");
  } else {
    const erwartet = GRUSSFORMELN[region];
    if (erwartet && mail.gruss.toLowerCase() !== erwartet.toLowerCase()) {
      melde(WARNUNG, "Grussformel",
        `'${mail.gruss}' passt nicht zur Region '${region}', erwartet: '${erwartet}'.`);
    }
  }
  for (const zeile of SIGNATUR) {
    if (!flach.includes(zeile.toLowerCase())) {
      melde(FEHLER, "Signatur", `Signaturzeile fehlt: '${zeile}'.`);
    }
  }

  // --- Sprachregister ----------------------------------------------------
  if (["ch", "fr", "it"].includes(region) && roh.includes("ß")) {
    melde(FEHLER, "Orthografie", "Schweizer Empfänger: ss statt ß.");
  }

  // --- Preise ------------------------------------------------------------
  for (const treffer of roh.matchAll(PREIS_MUSTER)) {
    melde(FEHLER, "Preise", `Preisangabe im Erstkontakt: '${treffer[0].trim()}'.`);
  }

  // --- Verfügbarkeit -----------------------------------------------------
  for (const satz of roh.split(/(?<=[.!?])\s+/)) {
    const treffer = satz.match(VERFUEGBARKEIT_MUSTER);
    if (treffer && !VERFUEGBARKEIT_ERLAUBT.test(satz)) {
      melde(FEHLER, "Verfügbarkeit",
        `Verfügbarkeit zugesagt ohne Freigabe: '${treffer[0].trim()}'.`);
    }
  }

  // --- KI-Marker ---------------------------------------------------------
  for (const marker of KI_MARKER) {
    if (flach.includes(marker)) {
      melde(FEHLER, "KI-Marker", `Blacklist-Formulierung: '${marker}'.`);
    }
  }

  // --- Superlative -------------------------------------------------------
  for (const treffer of roh.matchAll(SUPERLATIVE)) {
    melde(WARNUNG, "Superlativ",
      `Superlativ ohne Beleg: '${treffer[0]}'. Erlaubt ist nur die Tatsachenbehauptung 'als einziger Vermieter'.`);
  }

  // --- Formatierung ------------------------------------------------------
  if (GEDANKENSTRICH.test(roh)) {
    melde(FEHLER, "Formatierung", "Gedankenstriche sind als Stilmittel verboten.");
  }
  if (MARKDOWN_MUSTER.test(roh)) {
    melde(FEHLER, "Formatierung", "Kein Markdown im Mailtext (**, __, #).");
  }
  for (const zeile of mail.fliesstext.split("\n")) {
    if (BULLET_ZEILE.test(zeile)) {
      melde(FEHLER, "Formatierung",
        `Aufzählung im Mailtext, Kaltakquise ist Fliesstext: '${zeile.trim().slice(0, 40)}'.`);
      break;
    }
  }
  const emojis = [...new Set([...roh].filter(istEmoji))].sort();
  if (emojis.length) {
    melde(FEHLER, "Formatierung", `Emojis im Mailtext: ${emojis.join(" ")}.`);
  }

  // --- Normen ------------------------------------------------------------
  const voll = roh.match(NORM_IEC_VOLL);
  const kurz = roh.match(NORM_IEC_KURZ);
  const klasseA = roh.match(NORM_KLASSE_A);
  if (kurz && !voll) {
    melde(FEHLER, "Normen",
      "IEC 61000-4-30 beim ersten Nennen unvollständig, korrekt ist 'IEC 61000-4-30 Klasse A'.");
  }
  if (klasseA && voll && klasseA.index < voll.index) {
    melde(FEHLER, "Normen",
      "'Klasse A' steht vor der vollständigen Nennung von IEC 61000-4-30 Klasse A.");
  }
  if (klasseA && !kurz) {
    melde(FEHLER, "Normen",
      "'Klasse A' ohne Norm-Bezug, beim ersten Nennen IEC 61000-4-30 ausschreiben.");
  }

  // --- Einheiten ---------------------------------------------------------
  for (const treffer of roh.matchAll(EINHEIT_FALSCH)) {
    melde(FEHLER, "Einheiten",
      `Falsche Einheitenschreibweise: '${treffer[0]}' (kVA, kW, kWh, MVA sauber trennen).`);
  }

  // --- Fachkürzel --------------------------------------------------------
  for (const [kuerzel, aufloesung] of Object.entries(KUERZEL)) {
    if (new RegExp(`\\b${kuerzel}\\b`).test(roh) && !flach.includes(aufloesung.toLowerCase())) {
      melde(WARNUNG, "Fachkürzel",
        `'${kuerzel}' wird nicht aufgelöst, beim ersten Nennen '${aufloesung}' ausschreiben.`);
    }
  }

  // --- Call-to-Action ----------------------------------------------------
  if (mail.fliesstext) {
    if (!CTA_SIGNALE.test(mail.fliesstext)) {
      melde(FEHLER, "Call-to-Action",
        "Kein konkreter nächster Schritt erkennbar (Zeitfenster, Frage oder konkretes Angebot).");
    }
    const schwach = mail.fliesstext.match(CTA_SCHWACH);
    if (schwach) {
      melde(FEHLER, "Call-to-Action", `Weicher Abschluss statt konkretem Schritt: '${schwach[0]}'.`);
    }
  }

  // --- Lücken ------------------------------------------------------------
  for (const treffer of roh.matchAll(LUECKE_MUSTER)) {
    melde(HINWEIS, "Lücke", `Markierte Lücke vor dem Versand füllen: '${treffer[0]}'.`);
  }

  // --- Personalisierung --------------------------------------------------
  if (quelle && quelle.trim()) {
    const signale = signalwoerter(quelle);
    const getroffen = signale.filter((s) => flach.includes(s.toLowerCase()));
    if (signale.length && !getroffen.length) {
      melde(FEHLER, "Personalisierung",
        "Kein einziges Detail aus dem Input in der Mail. Diese Mail könnte an jede andere Firma gehen.");
    } else if (signale.length >= 4 && getroffen.length < 2) {
      melde(WARNUNG, "Personalisierung",
        `Nur ein Detail aus dem Input übernommen (${getroffen[0]}). Spezifischer werden.`);
    }
  }

  befunde.sort((a, b) => RANG[a.schwere] - RANG[b.schwere] || a.regel.localeCompare(b.regel, "de"));
  return befunde;
}

export const hatFehler = (befunde) => befunde.some((b) => b.schwere === FEHLER);

export function bilanz(befunde) {
  const z = { [FEHLER]: 0, [WARNUNG]: 0, [HINWEIS]: 0 };
  befunde.forEach((b) => (z[b.schwere] += 1));
  return z;
}

const VARIANTEN_KOPF = /^[ \t]*\**[ \t]*Variante[ \t]+([A-Z])\b[^\n]*$/gim;
const WARUM_ZEILE = /^\s*\*[^*].*wirkt.*\*\s*$/gim;
const ABSCHNITT_ENDE = /^\s*\**\s*(Follow-up|Annahme|Analyse)\b/im;

/** Schneidet aus der Modellausgabe die Blöcke 'Variante A/B' heraus. */
export function extrahiereVarianten(ausgabe) {
  const treffer = [...String(ausgabe).matchAll(VARIANTEN_KOPF)];
  if (!treffer.length) return [];

  return treffer.map((kopf, i) => {
    const von = kopf.index + kopf[0].length;
    const bis = i + 1 < treffer.length ? treffer[i + 1].index : ausgabe.length;
    let block = ausgabe.slice(von, bis);

    const schluss = block.match(ABSCHNITT_ENDE);
    if (schluss) block = block.slice(0, schluss.index);

    block = block
      .replace(WARUM_ZEILE, "")
      .replace(/^\s*```.*$/gm, "")
      .replace(/^\s*-{3,}\s*$/gm, "")
      .trim()
      .replace(/^"|"$/g, "");

    return { name: `Variante ${kopf[1].toUpperCase()}`, text: block.trim() };
  });
}

File: src/kaltakquise/pruefer.py
"""
Deterministischer Regel-Pruefer.

Prueft eine Mail gegen die harten Regeln des Systemprompts. Kein Modell,
kein Netz: gleiche Eingabe, gleiches Urteil. Damit ist "keine Fehler" eine
Zusicherung und keine Hoffnung.
"""
from __future__ import annotations

import re
import unicodedata
from dataclasses import dataclass

from . import regeln
from .mail import Mail, zerlege

FEHLER = "fehler"
WARNUNG = "warnung"
HINWEIS = "hinweis"

_RANG = {FEHLER: 0, WARNUNG: 1, HINWEIS: 2}


@dataclass(frozen=True)
class Befund:
    schwere: str
    regel: str
    text: str

    def __str__(self) -> str:  # pragma: no cover - reine Darstellung
        marke = {FEHLER: "FEHLER ", WARNUNG: "WARNUNG", HINWEIS: "HINWEIS"}[self.schwere]
        return f"[{marke}] {self.regel}: {self.text}"


def _normalisiert(text: str) -> str:
    """Kleinschreibung, Umlaute erhalten, Whitespace vereinheitlicht."""
    return re.sub(r"\s+", " ", text.lower())


def _ist_emoji(zeichen: str) -> bool:
    code = ord(zeichen)
    if unicodedata.category(zeichen) == "So":
        return True
    return any(start <= code <= ende for start, ende in regeln.EMOJI_BEREICHE)


def _signalwoerter(quelle: str) -> list[str]:
    """Eigennamen, Orte und Zahlen aus dem Input, an denen Personalisierung haengt."""
    signale: list[str] = []
    for wort in re.findall(r"\b[\wÄÖÜäöüß.-]{3,}\b", quelle):
        rein = wort.strip(".-")
        if not rein or rein.lower() in regeln.STOPP_WOERTER:
            continue
        if any(c.isdigit() for c in rein) or (rein[0].isupper() and len(rein) >= 4):
            signale.append(rein)
    # Reihenfolge erhalten, Duplikate entfernen
    return list(dict.fromkeys(signale))


def pruefe(
    text: str | Mail,
    *,
    region: str = "ch",
    quelle: str | None = None,
) -> list[Befund]:
    """Prueft eine Mail und liefert alle Befunde, schwerste zuerst."""
    mail = text if isinstance(text, Mail) else zerlege(text)
    befunde: list[Befund] = []

    def melde(schwere: str, regel: str, meldung: str) -> None:
        befunde.append(Befund(schwere, regel, meldung))

    roh = mail.roh
    flach = _normalisiert(roh)

    # --- Betreff ---------------------------------------------------------
    if not mail.betreff:
        melde(FEHLER, "Betreff", "Kein Betreff gefunden (erste Zeile 'Betreff: ...').")
    else:
        betreff_woerter = [w for w in mail.betreff.split() if w.strip()]
        if len(mail.betreff) > regeln.BETREFF_MAX_ZEICHEN:
            melde(
                WARNUNG,
                "Betreff",
                f"{len(mail.betreff)} Zeichen, maximal {regeln.BETREFF_MAX_ZEICHEN} "
                "(sonst auf dem Handy abgeschnitten).",
            )
        if not regeln.BETREFF_MIN_WOERTER <= len(betreff_woerter) <= regeln.BETREFF_MAX_WOERTER:
            melde(
                WARNUNG,
                "Betreff",
                f"{len(betreff_woerter)} Wörter, vorgesehen sind "
                f"{regeln.BETREFF_MIN_WOERTER} bis {regeln.BETREFF_MAX_WOERTER}.",
            )
        if "?" in mail.betreff or "!" in mail.betreff:
            melde(FEHLER, "Betreff", "Kein Frage- oder Ausrufezeichen im Betreff.")
        for verboten in regeln.BETREFF_VERBOTEN:
            if verboten in mail.betreff.lower():
                melde(FEHLER, "Betreff", f"Verbotenes Wort im Betreff: '{verboten}'.")

    # --- Anrede ----------------------------------------------------------
    if not mail.anrede:
        melde(FEHLER, "Anrede", "Keine Anrede gefunden.")
    if regeln.ANREDE_VERBOTEN in flach:
        melde(
            FEHLER,
            "Anrede",
            "'Sehr geehrte Damen und Herren' ist in der Kaltakquise verboten, "
            "Namen recherchieren oder Lücke markieren.",
        )

    # --- Einstieg --------------------------------------------------------
    erstes_wort = (mail.fliesstext.split() or [""])[0].strip(",.:;")
    if erstes_wort.lower() == "ich":
        melde(FEHLER, "Einstieg", "Der erste Satz darf nicht mit 'Ich' beginnen.")

    # --- Laenge ----------------------------------------------------------
    if not mail.fliesstext:
        melde(FEHLER, "Fliesstext", "Kein Fliesstext gefunden.")
    else:
        wortzahl = mail.wortzahl
        if wortzahl < regeln.FLIESSTEXT_MIN:
            melde(
                WARNUNG,
                "Länge",
                f"{wortzahl} Wörter Fliesstext, Ziel sind "
                f"{regeln.FLIESSTEXT_MIN} bis {regeln.FLIESSTEXT_MAX}.",
            )
        elif wortzahl > regeln.FLIESSTEXT_MAX:
            melde(
                FEHLER,
                "Länge",
                f"{wortzahl} Wörter Fliesstext, Maximum ist {regeln.FLIESSTEXT_MAX}. "
                "Kalt heisst kurz.",
            )

    # --- Grussformel und Signatur ----------------------------------------
    if not mail.gruss:
        melde(FEHLER, "Grussformel", "Keine Grussformel gefunden.")
    else:
        erwartet = regeln.GRUSSFORMELN.get(region)
        if erwartet and mail.gruss.lower() != erwartet.lower():
            melde(
                WARNUNG,
                "Grussformel",
                f"'{mail.gruss}' passt nicht zur Region '{region}', erwartet: '{erwartet}'.",
            )

    for zeile in regeln.SIGNATUR:
        if zeile.lower() not in flach:
            melde(FEHLER, "Signatur", f"Signaturzeile fehlt: '{zeile}'.")

    # --- Sprachregister --------------------------------------------------
    if region in ("ch", "fr", "it") and "ß" in roh:
        melde(FEHLER, "Orthografie", "Schweizer Empfänger: ss statt ß.")

    # --- Preise ----------------------------------------------------------
    for treffer in regeln.PREIS_MUSTER.finditer(roh):
        melde(
            FEHLER,
            "Preise",
            f"Preisangabe im Erstkontakt: '{treffer.group(0).strip()}'.",
        )

    # --- Verfuegbarkeit --------------------------------------------------
    for satz in re.split(r"(?<=[.!?])\s+", roh):
        treffer = regeln.VERFUEGBARKEIT_MUSTER.search(satz)
        if treffer and not regeln.VERFUEGBARKEIT_ERLAUBT.search(satz):
            melde(
                FEHLER,
                "Verfügbarkeit",
                f"Verfügbarkeit zugesagt ohne Freigabe: '{treffer.group(0).strip()}'.",
            )

    # --- KI-Marker -------------------------------------------------------
    for marker in regeln.KI_MARKER:
        if marker in flach:
            melde(FEHLER, "KI-Marker", f"Blacklist-Formulierung: '{marker}'.")

    # --- Superlative -----------------------------------------------------
    for treffer in regeln.SUPERLATIVE.finditer(roh):
        melde(
            WARNUNG,
            "Superlativ",
            f"Superlativ ohne Beleg: '{treffer.group(0)}'. "
            "Erlaubt ist nur die Tatsachenbehauptung 'als einziger Vermieter'.",
        )

    # --- Formatierung ----------------------------------------------------
    if regeln.GEDANKENSTRICH.search(roh):
        melde(FEHLER, "Formatierung", "Gedankenstriche sind als Stilmittel verboten.")
    if regeln.MARKDOWN_MUSTER.search(roh):
        melde(FEHLER, "Formatierung", "Kein Markdown im Mailtext (**, __, #).")
    for zeile in mail.fliesstext.split("\n"):
        if regeln.BULLET_ZEILE.match(zeile):
            melde(
                FEHLER,
                "Formatierung",
                f"Aufzählung im Mailtext, Kaltakquise ist Fliesstext: '{zeile.strip()[:40]}'.",
            )
            break
    emojis = sorted({z for z in roh if _ist_emoji(z)})
    if emojis:
        melde(FEHLER, "Formatierung", f"Emojis im Mailtext: {' '.join(emojis)}.")

    # --- Normen ----------------------------------------------------------
    voll = regeln.NORM_IEC_VOLL.search(roh)
    kurz = regeln.NORM_IEC_KURZ.search(roh)
    klasse_a = regeln.NORM_KLASSE_A.search(roh)
    if kurz and not voll:
        melde(
            FEHLER,
            "Normen",
            "IEC 61000-4-30 beim ersten Nennen unvollständig, korrekt ist "
            "'IEC 61000-4-30 Klasse A'.",
        )
    if klasse_a and voll and klasse_a.start() < voll.start():
        melde(
            FEHLER,
            "Normen",
            "'Klasse A' steht vor der vollständigen Nennung von IEC 61000-4-30 Klasse A.",
        )
    if klasse_a and not kurz:
        melde(
            FEHLER,
            "Normen",
            "'Klasse A' ohne Norm-Bezug, beim ersten Nennen IEC 61000-4-30 ausschreiben.",
        )

    # --- Einheiten -------------------------------------------------------
    for treffer in regeln.EINHEIT_FALSCH.finditer(roh):
        melde(
            FEHLER,
            "Einheiten",
            f"Falsche Einheitenschreibweise: '{treffer.group(0)}' "
            "(kVA, kW, kWh, MVA sauber trennen).",
        )

    # --- Fachkuerzel -----------------------------------------------------
    for kuerzel, aufloesung in regeln.KUERZEL.items():
        if re.search(rf"\b{kuerzel}\b", roh) and aufloesung.lower() not in flach:
            melde(
                WARNUNG,
                "Fachkürzel",
                f"'{kuerzel}' wird nicht aufgelöst, beim ersten Nennen "
                f"'{aufloesung}' ausschreiben.",
            )

    # --- Call-to-Action --------------------------------------------------
    if mail.fliesstext:
        if not regeln.CTA_SIGNALE.search(mail.fliesstext):
            melde(
                FEHLER,
                "Call-to-Action",
                "Kein konkreter nächster Schritt erkennbar (Zeitfenster, Frage "
                "oder konkretes Angebot).",
            )
        schwach = regeln.CTA_SCHWACH.search(mail.fliesstext)
        if schwach:
            melde(
                FEHLER,
                "Call-to-Action",
                f"Weicher Abschluss statt konkretem Schritt: '{schwach.group(0)}'.",
            )

    # --- Luecken ---------------------------------------------------------
    for treffer in regeln.LUECKE_MUSTER.finditer(roh):
        melde(
            HINWEIS,
            "Lücke",
            f"Markierte Lücke vor dem Versand füllen: '{treffer.group(0)}'.",
        )

    # --- Personalisierung ------------------------------------------------
    if quelle:
        signale = _signalwoerter(quelle)
        getroffen = [s for s in signale if s.lower() in flach]
        if signale and not getroffen:
            melde(
                FEHLER,
                "Personalisierung",
                "Kein einziges Detail aus dem Input in der Mail. Diese Mail könnte "
                "an jede andere Firma gehen.",
            )
        elif len(signale) >= 4 and len(getroffen) < 2:
            melde(
                WARNUNG,
                "Personalisierung",
                f"Nur ein Detail aus dem Input übernommen ({getroffen[0]}). "
                "Spezifischer werden.",
            )

    befunde.sort(key=lambda b: (_RANG[b.schwere], b.regel))
    return befunde


def hat_fehler(befunde: list[Befund]) -> bool:
    return any(b.schwere == FEHLER for b in befunde)


def bilanz(befunde: list[Befund]) -> str:
    zahl = {FEHLER: 0, WARNUNG: 0, HINWEIS: 0}
    for b in befunde:
        zahl[b.schwere] += 1
    return (
        f"{zahl[FEHLER]} Fehler, {zahl[WARNUNG]} Warnungen, {zahl[HINWEIS]} Hinweise"
    )
